File: platform/src/main/java/ru/abondin/hreasy/platform/sec/EmployeeUserContextMapperAdapter.java
package ru.abondin.hreasy.platform.sec;

import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.ldap.core.DirContextAdapter;
import org.springframework.ldap.core.DirContextOperations;
import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.ldap.userdetails.LdapUserDetailsMapper;
import org.springframework.security.ldap.userdetails.UserDetailsContextMapper;
import org.springframework.stereotype.Component;
import ru.abondin.hreasy.platform.BusinessError;
import ru.abondin.hreasy.platform.auth.AuthContext;
import ru.abondin.hreasy.platform.repo.employee.EmployeeAuthDomainService;

import java.time.Duration;
import java.util.Collection;

/**
 * Map LDAP record to employee information
 */
@Component
@RequiredArgsConstructor
@Slf4j
public class EmployeeUserContextMapperAdapter implements UserDetailsContextMapper {
    private final EmployeeAuthDomainService employeeAuthDomainService;
    private final Duration MAX_WAIT = Duration.ofSeconds(5);


    private UserDetailsContextMapper delegate = new LdapUserDetailsMapper();

    @Override
    public void mapUserToContext(UserDetails user, DirContextAdapter ctx) {
        delegate.mapUserToContext(user, ctx);
    }

    public UserDetails mapUserFromContext(DirContextOperations ctx, String username, Collection<? extends GrantedAuthority> authorities) {
        var user = delegate.mapUserFromContext(ctx, username, authorities);
        // FIXME One more block operation to support NOT reactive Spring API
        var employeeAuthInfoEntry = employeeAuthDomainService.findNotDismissedByUsername(username).block(MAX_WAIT);
        if (employeeAuthInfoEntry == null) {
            throw new BusinessError("errors.no.employee.found", username);
        }
        return new UserDetailsWithEmployeeInfo(user,
                employeeAuthInfoEntry.getId(),
                employeeAuthInfoEntry.getDepartmentId(),
                employeeAuthInfoEntry.getCurrentProjectId(),
                employeeAuthInfoEntry.getAccessibleDepartments(),
                employeeAuthInfoEntry.getAccessibleBas(),
                employeeAuthInfoEntry.getAccessibleProjects(),
                AuthContext.LoginType.LDAP.getValue());
    }
}

File: platform/src/main/kotlin/ru/abondin/hreasy/platform/sec/EmployeeUserContextMapperAdapter.kt
package ru.abondin.hreasy.platform.sec

import org.springframework.ldap.core.DirContextAdapter
import org.springframework.ldap.core.DirContextOperations
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.ldap.userdetails.LdapUserDetailsMapper
import org.springframework.security.ldap.userdetails.UserDetailsContextMapper
import org.springframework.stereotype.Component
import ru.abondin.hreasy.platform.repo.EmployeeRepo

/**
 * Add employeeId to the UserDetails
 */
@Component
class EmployeeUserContextMapperAdapter(private val employeeRepo: EmployeeRepo) : UserDetailsContextMapper {

    private val delegate: UserDetailsContextMapper = LdapUserDetailsMapper();

    override fun mapUserToContext(user: UserDetails?, ctx: DirContextAdapter?) {
        delegate.mapUserToContext(user, ctx);
    }

    override fun mapUserFromContext(ctx: DirContextOperations, username: String, authorities: MutableCollection<out GrantedAuthority>): UserDetails {
        val user = delegate.mapUserFromContext(ctx, username, authorities);
        val email = AuthHandler.emailFromUsername(username);
        // FIXME One more block operation to support NOT reactive Spring API
        val employeeId: Int? = employeeRepo.findIdByEmail(email).block();
        return UserDetailsWithEmployeeInfo(user, employeeId);
    }
}


data class UserDetailsWithEmployeeInfo(
        private val delegate: UserDetails,
        val employeeId: Int?
) : UserDetails {
    override fun getAuthorities(): MutableCollection<out GrantedAuthority> {
        return delegate.authorities;
    }

    override fun getPassword(): String {
        return delegate.password;
    }

    override fun getUsername(): String {
        return delegate.username
    }

    override fun isAccountNonExpired(): Boolean {
        return delegate.isAccountNonExpired;
    }

    override fun isAccountNonLocked(): Boolean {
        return delegate.isAccountNonLocked;
    }

    override fun isCredentialsNonExpired(): Boolean {
        return delegate.isCredentialsNonExpired;
    }

    override fun isEnabled(): Boolean {
        return delegate.isEnabled;
    }
}

